paimon-core/src/main/java/org/apache/paimon/utils/TagManager.java

```diff
@@ -119,6 +157,65 @@ public SortedMap<Snapshot, String> tags() {
         return tags;
     }
 
+    /** Rollback to tag. */
+    public void rollbackTo(String tagName, Snapshot latestSnapshot, TagDeletion tagDeletion) {
```

Comment:
Introduce a real `rollbackTo(long snapshotId)`, delete all tags and all snapshots bigger than it.